File: zaptorch/src/main/java/com/pyamsoft/zaptorch/dagger/service/TorchOffService.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.zaptorch.dagger.service;

import android.app.IntentService;
import android.content.Intent;
import com.pyamsoft.zaptorch.app.service.VolumeMonitorService;
import timber.log.Timber;

public class TorchOffService extends IntentService {

  public TorchOffService() {
    super(TorchOffService.class.getName());
  }

  @Override protected void onHandleIntent(Intent intent) {
    try {
      VolumeMonitorService.forceToggle();
    } catch (IllegalStateException e) {
      Timber.e(e, "onError");
    }
  }
}

File: zaptorch/src/main/java/com/pyamsoft/zaptorch/service/TorchOffService.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.zaptorch.service

import android.app.IntentService
import android.content.Intent
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import com.pyamsoft.pydroid.util.fakeBind
import com.pyamsoft.pydroid.util.fakeUnbind
import com.pyamsoft.zaptorch.Injector
import com.pyamsoft.zaptorch.ZapTorch
import com.pyamsoft.zaptorch.ZapTorchComponent
import timber.log.Timber

class TorchOffService : IntentService(TorchOffService::class.java.name), LifecycleOwner {

  private val registry = LifecycleRegistry(this)

  internal lateinit var presenter: TorchPresenter

  override fun getLifecycle(): Lifecycle {
    return registry
  }

  override fun onCreate() {
    super.onCreate()
    Injector.obtain<ZapTorchComponent>(applicationContext)
        .inject(this)

    registry.fakeBind()
  }

  override fun onDestroy() {
    super.onDestroy()
    ZapTorch.getRefWatcher(this)
        .watch(this)

    registry.fakeUnbind()
  }

  override fun onHandleIntent(intent: Intent?) {
    try {
      presenter.toggle()
    } catch (e: IllegalStateException) {
      Timber.e(e, "onError")
    }
  }
}
